bastionzero/bzero
# SignalR: Send ping messages from client + handle server pings

## Description of the change

Adds both client ping messages and server ping timeout to our signalR client. This fixes a variety of issues in regard to detecting when a websocket connection is broken and being able to properly recover in these cases.

+ Adds client pings at regular intervals (every 15s) if there are no other messages sent in the websocket.  The [asp .net core signalR server implementation](https://github.com/dotnet/aspnetcore/blob/34a16dcf5043b4f02756e9d0727943c385cd2810/src/SignalR/docs/specs/HubProtocol.md) has a client timeout where it will close the connection if it doesnt receive any messages from the client. This timeout defaults to 30s, however, in order to trigger this timeout **the client must send at least one ping message**. Because our client was never sending any ping messages the server would never timeout our client connections when they were broken.

+ Adds server ping timeout to the client. Similarly, the server will send pings to the client at regular intervals. It is [recommended](https://github.com/dotnet/aspnetcore/blob/34a16dcf5043b4f02756e9d0727943c385cd2810/src/SignalR/docs/specs/HubProtocol.md#ping-aka-keep-alive) that the client be listening for these pings and also timing out the connection if it doesnt receive a ping within a factor of server the ping interval (also 15s). This helps the client detect when its websocket connection is broken - without this, because we dont have any read timeout at the underlying gorilla websocket transport layer, we do not detect that the websocket connection is broken until a much longer timeout on the tcp connection itself.

## Testing

Testing recovery on network connectivity issues:

Dependent changes:

backend: https://github.com/bastionzero/webshell-backend/pull/1552
agent: https://github.com/bastionzero/bzero/pull/229. This is needed so the agent wont close its connection 1m after the daemon disconnects and instead will use the much longer IdleTimeout (7days).

1. First of make sure agent and zli and backend are updated.
2.  zli connect <target> - this fix should apply to any connection type (kube/web/db/shell/ssh).
3. Turn off wifi
4. Server should timeout the close connection after 30s and client should also timeout and close signalR client after 30s which triggers entering a reconnect backoff loop in the dataconnection component.
5. Turn wifi back on. The client should reconnect
6. Connection should be working again and client should be able to send/receive messages from the agent. Note: there is definitely still an issue for shell/ssh in our datachannel pipelining code that prevents the datachannel from being used even after the underlying websocket reconnects. This is a separate issue and will likely be fixed in a subsequent PR.

**backend branch:** 
**zli branch:** 
**charts branch:** 

#### Ready to run system tests?


## Relevant release note information

Release Notes: Implements client ping protocol at the signalR layer to detect when the websocket is broken. Also fixes a few bugs at the data connection layer that prevented the daemon/agent from working properly after one party disconnects and reconnects.   

## Related JIRA tickets

Relates to JIRA: CWC-2486

## Have you considered the security impacts?

Does this PR have any security impact?


If yes, please explain:
